jwst/msaflagopen/msaflag_open.py

```diff
@@ -0,0 +1,235 @@
+        science data with DQ flags of affected modified
+
+    """
+    #
+    # Use the machinery in assign_wcs to create a WCS object for the bad shutters
+    pipeline = slitlets_wcs(input_datamodel, wcs_refnames, failed_slitlets)
+    wcs = WCS(pipeline)
+    # Create output as a copy of the input science data model so we can overwrite
+    # the wcs with the wcs for the failed open shutters
+    temporary_copy = input_datamodel.copy()
+    temporary_copy.meta.wcs = wcs
+    dq_array = input_datamodel.dq
+    for slitlet in failed_slitlets:
+        #
+        # Pick the WCS for this slitlet from the WCS of the exposure
+        thiswcs = nrs_wcs_set_input(temporary_copy, slitlet.name)
+        #
+        # Convert the bounding box for this slitlet to a set of indices to use as a slice
+        xmin, xmax, ymin, ymax = boundingbox_to_indices(temporary_copy,
+                                                        thiswcs.bounding_box)
+        #
+        # Make a grid of points within the slice
+        y_indices, x_indices = np.mgrid[ymin:ymax, xmin:xmax]
+        #
+        # Calculate the arrays of coordinates for each pixel in the slice
+        coordinate_array = thiswcs(x_indices, y_indices)
+        #
+        # The coordinate_array is a tuple of arrays, one for each output coordinate
+        # In this case there should be 3 arrays, one each for RA, Dec and Wavelength
+        # For pixels outside the slitlet, the arrays have NaN
+        #
+        # Make a subarray from these coordinate arrays by setting pixels that aren't
+        # NaN to FAILDOPENFLAG, the rest to 0
+        dq_subarray = wcs_to_dq(coordinate_array, FAILEDOPENFLAG)
+        #
+        # bitwise-or this subarray with the slice in the original exposure's DQ array
+        dq_array = or_subarray_with_array(dq_array, dq_subarray, xmin, xmax, ymin, ymax)
+    #
+    # Set the dq array of the input datamodel to the corrected dq array
+    input_datamodel.dq = dq_array
+    return input_datamodel
+
+def boundingbox_to_indices(data_model, bounding_box):
+    """
+    Takes a bounding_box (tuple of tuples: ((x1, x2), (y1, y2)) and
+    a datamodel and calculates the range of indices in the X and Y dimensions
+    of the overlap between the bounding box and the datamodel's data array.
+
+    Parameters
+    ----------
+    data_model: data model object
+        the input science datamodel
+
+    bounding_box: Tuple of tuples
+        Bounding box returned from wcs object
+
+    Returns
+    -------
+    xmin, xmax, ymin, ymax: integers
+        Range of indices of overlap between science data array and bounding box
+
+    """
+    nrows, ncols = data_model.data.shape
+    ((x1, x2), (y1, y2)) = bounding_box
+    xmin = int(min(x1, x2))
+    xmin = max(xmin, 0)
+    xmax = int(max(x1, x2)) + 1
+    xmax = min(xmax, ncols)
+    ymin = int(min(y1, y2))
+    ymin = max(ymin, 0)
+    ymax = int(max(y1, y2)) + 1
+    ymax = min(ymax, nrows)
+    return (xmin, xmax, ymin, ymax)
```

Comment:
data is in /user/rij/jwst/test/test_badflag.  Dataset is jw00038001001_01101_00001_NRS1_assignwcsstep.fits, file with bad shutters is the default msaoper reference file: jwst_nirspec_msaoper_0001.json
Here are the bounding boxes for the 20 shutters that have 'open' as status:
((2422.5396705052904, 2047.0), (1917.2985352895864, 1939.5456127818488))
((2143.6896349047074, 2047.0), (1834.28258973113, 1852.416720048566))
((1774.8616957418017, 2047.0), (1641.4066319191058, 1654.8130840329404))
((1725.9789471362044, 2047.0), (1331.2521667305348, 1343.9115576438633))
((1633.1676372217519, 2047.0), (1601.8171259529258, 1613.8900744506313))
((2340.2485270529191, 2047.0), (689.01367783442879, 706.88469883647952))
((1626.7062562097904, 2047.0), (248.12045790047631, 259.31998557812341))
((1606.1555962184043, 2047.0), (172.35678755141956, 183.39298857156538))
((1208.1238113799989, 2047.0), (1271.4489883888211, 1282.5491699711795))
((1206.9090311714133, 2047.0), (1241.3647483447105, 1252.4459469368044))
((974.2175417492407, 2047.0), (1835.8351743211965, 1849.6870108103301))
((915.41673598361172, 2047.0), (1260.4479615661339, 1273.9267488544233))
((715.52445748929006, 2047.0), (1699.5255694894374, 1716.0949338212693))
((607.22254419112255, 1939.8850639079114), (1638.8442151027577, 1656.6150881542635))
((479.82133427274897, 1810.9081350672436), (1383.0136641385782, 1401.6157074135763))
((497.28004521715195, 1831.027783841872), (1867.5806462679666, 1887.4655951038583))
((478.44043915921043, 1810.244216709387), (1523.0325156706551, 1542.0689789804369))
((406.56494046640756, 1732.8820368485615), (209.91766590473725, 226.45149763960018))
((375.57632565268511, 1703.2269016065877), (628.08322699577081, 645.8179767893497))
((366.43540534963176, 1694.1822803799391), (658.21419719554922, 676.11543190835039))
It's the first 2 that cause a problem...